byoeb-integrations: write tests for embeddings and llms

mamuqsit opened the pull request:
## Introduction
This PR adds remaining tests for constructor errors to ensure an embedding or an LLM can not be in an invalid state when successfully instantiated (i.e., `get_embedding_function()` / `get_llm_client()` does not return None). Mocker fixtures have been implemented for existing LLMs to make subsequent tests easier. In test_agenerate_response (testing parallelism), errors from child now propagate to main.

## Coverage improvement
- [byoeb_integrations/embeddings/chroma/llama_index_azure_openai.py](https://github.com/A4i-tech/byoeb/blob/a4i/main/byoeb-v1/byoeb-integrations/byoeb_integrations/embeddings/chroma/llama_index_azure_openai.py) (90% -> 100%)
- [byoeb_integrations/embeddings/llama_index/azure_openai.py](https://github.com/A4i-tech/byoeb/blob/a4i/main/byoeb-v1/byoeb-integrations/byoeb_integrations/embeddings/llama_index/azure_openai.py) (60% -> 100%)
- [byoeb_integrations/llms/azure_openai/async_azure_openai.py](https://github.com/A4i-tech/byoeb/blob/a4i/main/byoeb-v1/byoeb-integrations/byoeb_integrations/llms/azure_openai/async_azure_openai.py) (64% -> 96%)
- [byoeb_integrations/llms/llama_index/llama_index_azure_openai.py](https://github.com/A4i-tech/byoeb/blob/a4i/main/byoeb-v1/byoeb-integrations/byoeb_integrations/llms/llama_index/llama_index_azure_openai.py) (49% -> 98%)
- [byoeb_integrations/llms/llama_index/llama_index_openai.py](https://github.com/A4i-tech/byoeb/blob/a4i/main/byoeb-v1/byoeb-integrations/byoeb_integrations/llms/llama_index/llama_index_openai.py) (33% -> 98%)

mamuqsit committed:
Add tests for AzureOpenAIEmbed
- Tests error handling during bad combination of ctor params
- Tests get_embedding_function() is not None after successful instantiation
Parameter values were ported from test_embeddings_chroma.

mamuqsit committed:
Fix AzureOpenAIEmbeddingFunction.__call__ not highlighted in coverage
Coverage cant mark code replaced by patching __call__, so we mock the dependency instead and execute the actual __call__ implementation.

mamuqsit committed:
llms: extract common test cases to reduce boilerplate

mamuqsit committed:
openai-llms: run instantiation and prompt generation tests on a mocked llm

mamuqsit committed:
azure-openai-llms: run instantiation, prompt generation tests on mocked llm instances

mamuqsit committed:
Cleanup unused imports

mamuqsit committed:
test-llms-openai: use hardcoded values instead of env vars
These environment variables are not available in workflow.

N1kunj1998 commented:
test inside byoeb_integrations/llms/tests/test_llms_openai.py this file is failing, can you check this error

mamuqsit commented:
> test inside byoeb_integrations/llms/tests/test_llms_openai.py this file is failing, can you check this error

Yep, the environment variables aren't available in test environment, I'll substitute them.

mamuqsit committed:
Apply test_llms_generic tests only on llamaindex instances
LLMs from llamaindex have similar message structure as openai's (i.e., {role, content}) but are differently typed (openai ChatCompletionMessage vs. llamaindex ChatMessage). This test cannot be extended to async_azure_openai.

mamuqsit committed:
agenerate_response: extend the test to both - openai and llamaindex instances

mamuqsit committed:
test_llms_azure_openai: remove outdated get_llm_client test for openai llm

mamuqsit committed:
llama_simple: test AsyncCompletions.create thru mocking instead

mamuqsit committed:
test_agenerate_response: collect errors from child and raise them after join()
Errors in child thread do not propagate to main thread and would be left unreported

mamuqsit commented:
Right, I omitted the follow up instruction from the PR. Delegated LLMs are vendored from different libraries (openai and llamaindex) which means implementation varies and the failing test can not be generalized. Went with mocking AsyncCompletions.create directly for openai-vendored LLMs.

Ready to merge.

mamuqsit committed:
async_azure_openai: cleanup accidental modification

N1kunj1998 reviewed (approved):
Changes looks good to me
